File: app/src/main/java/com/dougkeen/bart/controls/CheckableLinearLayout.java
package com.dougkeen.bart.controls;

import android.content.Context;
import androidx.appcompat.widget.LinearLayoutCompat;
import android.util.AttributeSet;
import android.widget.Checkable;

/** A layout that exposes its checked state to a state-list background. */
public class CheckableLinearLayout extends LinearLayoutCompat implements Checkable {

    private boolean mChecked;
    private static final int[] CHECKED_STATE = {android.R.attr.state_checked};

    public CheckableLinearLayout(Context context) {
        super(context);
    }

    public CheckableLinearLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public CheckableLinearLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public boolean isChecked() {
        return mChecked;
    }

    @Override
    public void setChecked(boolean checked) {
        mChecked = checked;
        refreshDrawableState();
    }

    @Override
    public void toggle() {
        setChecked(!isChecked());
    }

    @Override
    protected int[] onCreateDrawableState(int extraSpace) {
        int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
        if (mChecked) {
            mergeDrawableStates(drawableState, CHECKED_STATE);
        }
        return drawableState;
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/controls/CheckableLinearLayout.kt
package com.dougkeen.bart.controls

import android.content.Context
import android.util.AttributeSet
import android.widget.Checkable
import androidx.appcompat.widget.LinearLayoutCompat

/** A layout that exposes its checked state to a state-list background. */
class CheckableLinearLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : LinearLayoutCompat(context, attrs, defStyleAttr), Checkable {
    private var checked = false

    override fun isChecked(): Boolean = checked

    override fun setChecked(checked: Boolean) {
        this.checked = checked
        refreshDrawableState()
    }

    override fun toggle() {
        isChecked = !isChecked
    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val drawableState = super.onCreateDrawableState(extraSpace + 1)
        if (checked) {
            mergeDrawableStates(drawableState, CHECKED_STATE)
        }
        return drawableState
    }

    private companion object {
        val CHECKED_STATE = intArrayOf(android.R.attr.state_checked)
    }
}
